[xbmc][win32] First step in supporting VS2017

Paxxi opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
<!--- Describe your change in detail -->
Adds vswhere as a package.

Will try to find vs2017 and fall back to vs2015 in case it's not found.
SDK version is still hard coded and pinned to 14393 to avoid any breakage on jenkins.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

afedchin commented on tools/buildsteps/windows/BuildSetup.bat:
```diff
@@ -1,319 +1,319 @@
+WEBSITE
+
+FOR %%p IN (%version_props%) DO (
+  FOR /f "delims=" %%v IN ('%sed_exe% -n "/%%p/ s/%%p *//p" %base_dir%\version.txt') DO SET %%p=%%v
```
`^M` shouldn't be here

Paxxi replied:
yeah I think it got a bit messed up because I had turned off autocrlf in git

Paxxi replied:
hmm I've squashed it and looked at the end of line characters for everything and this line has the same crlf as every other line but for some reason github shows the ^M character.

afedchin reviewed (approved):
looks good in general, but something wrong with BuildSetup.bat

Rechi commented:
@Paxxi fix the commit messages and I'll review it afterwards

Paxxi commented:
@Rechi squashed it now and gotten rid of all the nonsense commits

Rechi commented:
whole BuildSetup.bat file still has a line ending change

Paxxi commented:
ok third time's the charm :)

Rechi commented:
The 64 bit toolchains should still be the preferred ones, but with your changes they aren't.
WIN-UWP-ARM fails, because you have to call vcvarsall.bat with `x86_arm` or `amd64_arm` if you run it on a non ARM OS.

Paxxi commented:
ahh that's the issue with arm. I'll look at fixing those bits.
Does it look ok otherwise?

MartijnKaijser commented:
jenkins build this please

MartijnKaijser commented:
jenkins build this please

Paxxi commented:
So here's the plan for the upgrade to vs2017.

1. Get this merged
2. Install vs2017 and sdk ver 10.0.16299.0 on all build machines
3. PR a bump to the sdk version
4. uninstall vs2015 and sdk 10.0.14393.0 on all build machines
5. profit

afedchin commented:
should we update vcredist?

Paxxi commented:
forgot about vcredist. Not sure if we need to but makes sense to bump it as well to match.

Rechi commented on .gitignore:
```diff
@@ -58,7 +59,7 @@ testMain
 
 # osx specifics
-/build/
+/build*/
```
should go into a separate commit, if you want to keep it

Rechi commented on tools/buildsteps/windows/BuildSetup.bat:
```diff
@@ -1,5 +1,5 @@
 @ECHO OFF
-SETLOCAL ENABLEDELAYEDEXPANSION
+REM SETLOCAL ENABLEDELAYEDEXPANSION
```
If not needed remove it, instead commenting it

Rechi commented on tools/buildsteps/windows/vswhere.bat:
```diff
@@ -0,0 +1,63 @@
+SET toolsdir=%arch%
+
+IF "%arch%" NEQ "x64" (
+  SET vcarch=%vcarch%_%arch%
```
We should try with `amd64_%arch%` first and if that fails fall back to `%arch%` like we currently do it.

Paxxi replied:
do we really care about building on 32-bit versions?

Rechi replied:
Will the `amd64_%arch%` version always be available if `%arch% is available?
If yes I don't care much, otherwise it has to be adjusted.

Paxxi replied:
Should be. There's no separate option to install one or the other
![image](https://user-images.githubusercontent.com/874927/34469302-4b6cbd36-ef1c-11e7-8319-98b9ddd01207.png)

Rechi commented on tools/buildsteps/windows/vswhere.bat:
```diff
@@ -0,0 +1,63 @@
+  FOR /f "usebackq tokens=1* delims=" %%i in (`%vswhere% -legacy -property installationPath`) do (
+    IF EXIST "%%i\VC\vcvarsall.bat" (
+      SET vcvars="%%i\VC\vcvarsall.bat"
+      SET vsver=14
```
exit for loop after first existing path

Paxxi replied:
it's doesn't really matter much as for most systems we'll only get a single result back

Rechi replied:
On my system I get two results back, although for the second one (12.0) the if check returns false but others may have installed older versions of VS.

Paxxi replied:
I was thinking of the wrong check. This check will be removed after we bump the sdk, it's only there currently to not break jenkins so we can do the upgrade in steps.

Also it seems harder than it should be to break out of a for loop see https://stackoverflow.com/questions/6728798/exiting-out-of-a-for-loop-in-a-batch-file haven't tested it myself though.

Rechi commented on tools/buildsteps/windows/win32-uwp/make-addons.bat:
```diff
@@ -1,8 +1,6 @@
-@ECHO OFF
```
line endings changed in whole file

Paxxi replied:
Not entirely sure how to fix that. I've since switched my checked out repo to autocrlf=true and cloning a new version without it checks it out as LF file endings.

Rechi reviewed (commented):
Should we already use VS2017 as preferred version and use VS2015 as fallback?

Paxxi commented:
I think we should as this is intended to upgrade the vs2017 and build hosts don't have it installed it's a safe setting imho.

Dropped the gitignore change and remove the commented out setlocal. Seem to have fixed the line endings but I'm never really sure  😄

Paxxi commented:
jenkins build this please

MartijnKaijser commented:
jenkins build this please

MartijnKaijser commented:
jenkins build this please

Rechi commented:
has to be squashed

MartijnKaijser commented:
jenkins build this please

Paxxi committed:
Adds vswhere as a package.

Will try to find vs2017 and fall back to vs2015 in case it's not found.
SDK version is still hard coded and pinned to 14393 to avoid any breakage on jenkins.
